PGIP/Server.hs

```diff
@@ -1806,27 +1822,224 @@ consNode :: LibEnv -> LibName -> DGraph -> (Int, DGNodeLab)
+                    AbsState.ConsChecker cc, c, Nothing)])
+
+findConsChecker :: Maybe String -> G_sublogics -> Maybe String
+                -> IO (G_cons_checker, AnyComorphism)
+findConsChecker translationM gSublogic consCheckerNameM = do
+  consList <- getFilteredConsCheckers translationM gSublogic
   let findCC x = filter ((== x ) . getCcName . fst) consList
-      mcc = maybe (findCC "darwin") findCC mp
-  case mcc of
+      consCheckersL = maybe (findCC "darwin") findCC consCheckerNameM
+  case consCheckersL of
         [] -> fail "no cons checker found"
-        ((cc, c) : _) -> lift $ do
-          cstat <- consistencyCheck useTh cc c ln le dg nl $ fromMaybe 1 tl
-          -- Consistency Results are stored in LibEnv via DGChange object
-          let cSt = sType cstat
-              le'' = if cSt == CSUnchecked then le else
-                     Map.insert ln (changeDGH dg $ SetNodeLab lb
-                       (i, case cSt of
-                             CSInconsistent -> markNodeInconsistent "" lb
-                             CSConsistent -> markNodeConsistent "" lb
-                             _ -> lb)) le
-          return (le'', [(" ", drop 2 $ show cSt, show cstat,
-                          AbsState.ConsChecker cc, c, Nothing)])
-
-proveNode :: LibEnv -> LibName -> DGraph -> (Int, DGNodeLab) -> G_theory
-  -> G_sublogics -> Bool -> Maybe String -> Maybe String -> Maybe Int
-  -> [String] -> [String] -> ResultT IO (LibEnv, [ProofResult])
+        (gConsChecker, comorphism) : _ -> return (gConsChecker, comorphism)
+
+reasonREST :: HetcatsOpts -> LibEnv -> LibName -> DGraph -> ProverMode
+           -> String -> ReasoningParameters
+           -> ResultT IO (LibEnv, [(String, [ProofResult])])
+reasonREST opts libEnv libName dGraph_ proverMode location reasoningParameters = do
+  reasoningCacheE <- liftIO $ buildReasoningCache
+  failOnLefts reasoningCacheE
+  let reasoningCache1 = map unRight $ filter (not . isLeft) reasoningCacheE
+  reasoningCache2 <- liftIO $ PGIP.Server.setupReasoning opts reasoningCache1
+  (libEnv', cacheGoalsAndProofResults) <-
+    PGIP.Server.performReasoning opts libEnv libName
+      dGraph_ location reasoningCache2
+  let nodesAndProofResults = map
+        (\ (nodeLabel, proofResults) ->
+          (show (getName $ dgn_name nodeLabel), proofResults)
+        )
+        cacheGoalsAndProofResults
+  return (libEnv', nodesAndProofResults)
+  where
+    useDatabase :: Bool
+    useDatabase = format reasoningParameters == Just "db"
+
+    failOnLefts :: ReasoningCacheE -> ResultT IO ()
+    failOnLefts reasoningCache =
+      let lefts_ = filter isLeft reasoningCache
+      in  if null lefts_
+          then return ()
+          else fail $ unlines $ map (\ (Left message_) -> message_) lefts_
+
+    isLeft :: Either a b -> Bool
+    isLeft (Left _) = True
+    isLeft _ = False
+
+    unRight :: Either a b -> b
+    unRight (Right x) = x
+    unRight _ = error "PGIP.Server.unRight received a Left."
```

Comment:
shorter: `unRight = fromRight $ error "PGIP.Server.unRight received a Left."`
Perhaps you could make the error message a bit more informative?

Reply:
I cannot use `fromRight`. It could not be found even though I have imported `Data.Either`.